pytensor/link/numba/dispatch/elemwise.py

```diff
@@ -539,44 +539,54 @@ def numba_funcify_DimShuffle(op, **kwargs):
 
     ndim_new_shape = len(shuffle) + len(augment)
 
+    no_transpose = all(i == j for i, j in enumerate(transposition))
+    if no_transpose:
+
+        @numba_basic.numba_njit
+        def transpose(x):
+            return x
+
+    else:
+
+        @numba_basic.numba_njit
+        def transpose(x):
+            return np.transpose(x, transposition)
+
+    shape_template = (1,) * ndim_new_shape
+
+    # When `len(shuffle) == 0`, the `shuffle_shape[j]` expression above is
+    # is typed as `getitem(Tuple(), int)`, which has no implementation
+    # (since getting an item from an empty sequence doesn't make sense).
+    # To avoid this compile-time error, we omit the expression altogether.
     if len(shuffle) > 0:
 
         @numba_basic.numba_njit
-        def populate_new_shape(i, j, new_shape, shuffle_shape):
-            if i in augment:
-                new_shape = numba_basic.tuple_setitem(new_shape, i, 1)
-                return j, new_shape
-            else:
-                new_shape = numba_basic.tuple_setitem(new_shape, i, shuffle_shape[j])
-                return j + 1, new_shape
+        def find_shape(array_shape):
```

Comment:
Should we shortcut when the output static shape is known?

```python
pt.random.normal(size=(2, 3)).dimshuffle(1, 0).type.shape  # (3, 2)
```

Reply:
I guess that might save a tiny bit of runtime and/or compile time, but if that were incorrect for some reason we would end up writing to arrays out of bounds. There is no error checking after this point.
I think it is safer to always infer the output shape from the inputs...

Reply:
In that case the graph would be inconsistent and should fail anyway.

Reply:
Yes, it *should* fail. But if we just assume the shapes are correct, we might just silently corrupt memory and *not* fail in an obvious way. :-)

Reply:
ah, sorry I somehow though this was the shape code for the llvm-elemwise...
Here the reshape should just fail if we provide something incorrect. So yes, I think we can use extra info we have here, I'll update the PR.

Reply:
Either way it's probably a minor performance difference. I was more thinking out loud in what types of place can we benefit from static shape info.